{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "geneToLength = {}\n",
    "genes = pd.read_csv('/data/khandekara2/bed_CpGs/mart_export.bed', sep='\\t', header=None)\n",
    "for gene, length in zip(genes.iloc[:,5], genes.iloc[:,3]):\n",
    "    geneToLength[gene] = length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1322\n"
     ]
    }
   ],
   "source": [
    "print (geneToLength['ACBD7'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#using the MALY cancer data and it's corresponding benign tissue data, \n",
    "#find those genes whose methylation levels in coding regions departed significantly(p <= 0.05) from the cds average \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "df = pd.read_csv('P_cg_results_highly_methylated.tsv', sep='\\t')#normal tissue\n",
    "df2 = pd.read_csv('P_cg_results_lowly_methylated.tsv', sep='\\t')#normal tissue\n",
    "maly = pd.read_csv('P_cg_results_MALY_highly_methylated.tsv', sep='\\t')#tumor\n",
    "maly2 = pd.read_csv('P_cg_results_MALY_lowly_methylated.tsv', sep='\\t')#tumor\n",
    "thymus = df[df['tissue'] == 'Thymus']\n",
    "thymus2 = df2[df2['tissue'] == 'Thymus']\n",
    "thymus = thymus.dropna()\n",
    "thymus2 = thymus2.dropna()\n",
    "thymus2 = thymus2.filter(['p_value'], axis=1)\n",
    "maly = maly.dropna()\n",
    "maly2 = maly2.dropna()\n",
    "maly2 = maly2.filter(['p_value'], axis=1)\n",
    "thymus2.rename(columns={'p_value': 'p_value_low'}, inplace=True)\n",
    "maly2.rename(columns={'p_value': 'p_value_low'}, inplace=True)\n",
    "maly['p_value_low'] = maly2['p_value_low']\n",
    "thymus['p_value_low'] = thymus2['p_value_low']\n",
    "thymus['high_benign'] = np.where(thymus['p_value'] <= 0.05, 'yes', 'no')\n",
    "thymus['baseline_benign'] = np.where((thymus['p_value'] > 0.05) & (thymus['p_value_low'] > 0.05), 'yes', 'no')\n",
    "thymus['low_benign'] = np.where(thymus['p_value_low'] <= 0.05, 'yes', 'no')\n",
    "maly['low'] = np.where(maly['p_value_low'] <= 0.05, 'yes', 'no')\n",
    "maly['baseline'] = np.where((maly['p_value'] > 0.05) & (maly['p_value_low'] > 0.05), 'yes', 'no')\n",
    "maly['high'] = np.where(maly['p_value'] <= 0.05, 'yes', 'no')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>gene</th>\n",
       "      <th>total</th>\n",
       "      <th>methylated</th>\n",
       "      <th>p_value</th>\n",
       "      <th>prior</th>\n",
       "      <th>length</th>\n",
       "      <th>p_value_low</th>\n",
       "      <th>low</th>\n",
       "      <th>baseline</th>\n",
       "      <th>high</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>CCDC24</td>\n",
       "      <td>31</td>\n",
       "      <td>14</td>\n",
       "      <td>0.999309</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1463</td>\n",
       "      <td>0.002283</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>C1orf50</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.71</td>\n",
       "      <td>996</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>PPFIA4</td>\n",
       "      <td>79</td>\n",
       "      <td>55</td>\n",
       "      <td>0.658577</td>\n",
       "      <td>0.71</td>\n",
       "      <td>5045</td>\n",
       "      <td>0.435182</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>CAP1</td>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "      <td>0.132087</td>\n",
       "      <td>0.71</td>\n",
       "      <td>3105</td>\n",
       "      <td>0.947553</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>MMACHC</td>\n",
       "      <td>27</td>\n",
       "      <td>21</td>\n",
       "      <td>0.294013</td>\n",
       "      <td>0.71</td>\n",
       "      <td>2981</td>\n",
       "      <td>0.838458</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tissue     gene  total  methylated   p_value  prior  length  p_value_low  \\\n",
       "0  Thymus   CCDC24     31          14  0.999309   0.71    1463     0.002283   \n",
       "2  Thymus  C1orf50     21           4  1.000000   0.71     996     0.000001   \n",
       "3  Thymus   PPFIA4     79          55  0.658577   0.71    5045     0.435182   \n",
       "4  Thymus     CAP1     24          20  0.132087   0.71    3105     0.947553   \n",
       "5  Thymus   MMACHC     27          21  0.294013   0.71    2981     0.838458   \n",
       "\n",
       "   low baseline high  \n",
       "0  yes       no   no  \n",
       "2  yes       no   no  \n",
       "3   no      yes   no  \n",
       "4   no      yes   no  \n",
       "5   no      yes   no  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maly.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8856\n",
      "4470\n",
      "4898\n"
     ]
    }
   ],
   "source": [
    "#print statistics for benign tissue(thymus)...how many genes were in each category\n",
    "high_benign = [e for e in thymus['high_benign'] if e == 'yes']\n",
    "low_benign = [e for e in thymus['low_benign'] if e == 'yes']\n",
    "baseline_benign = [e for e in thymus['baseline_benign'] if e == 'yes']\n",
    "print (len(high_benign))\n",
    "print (len(low_benign))\n",
    "print (len(baseline_benign))\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6329\n",
      "4680\n",
      "6939\n"
     ]
    }
   ],
   "source": [
    "#print statistics for cancer samples(MALY)...how many genes were in each category\n",
    "high_maly = [e for e in maly['high'] if e == 'yes']\n",
    "low_maly = [e for e in maly['low'] if e == 'yes']\n",
    "baseline_maly = [e for e in maly['baseline'] if e == 'yes']\n",
    "print (len(high_maly))\n",
    "print (len(low_maly))\n",
    "print (len(baseline_maly))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#merge benign and cancer dataframes on gene name so that we have all information for a gene in one table and write out to tsv file\n",
    "maly = maly.drop('tissue', axis=1)\n",
    "merged = pd.merge(thymus, maly, on='gene', how='inner')\n",
    "merged['p_value_y'] = merged['p_value_y'].apply(lambda x: round(x, 2))\n",
    "merged.head()\n",
    "merged.to_csv('maly_cds_differential_methylation.tsv', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>gene</th>\n",
       "      <th>total_x</th>\n",
       "      <th>methylated_x</th>\n",
       "      <th>p_value_x</th>\n",
       "      <th>prior_x</th>\n",
       "      <th>length_x</th>\n",
       "      <th>p_value_low_x</th>\n",
       "      <th>high_benign</th>\n",
       "      <th>baseline_benign</th>\n",
       "      <th>low_benign</th>\n",
       "      <th>total_y</th>\n",
       "      <th>methylated_y</th>\n",
       "      <th>p_value_y</th>\n",
       "      <th>prior_y</th>\n",
       "      <th>length_y</th>\n",
       "      <th>p_value_low_y</th>\n",
       "      <th>low</th>\n",
       "      <th>baseline</th>\n",
       "      <th>high</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>130</td>\n",
       "      <td>83</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.74</td>\n",
       "      <td>1722</td>\n",
       "      <td>0.01</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>108</td>\n",
       "      <td>76</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1722</td>\n",
       "      <td>0.478854</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.74</td>\n",
       "      <td>9221</td>\n",
       "      <td>1.00</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>24</td>\n",
       "      <td>22</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.71</td>\n",
       "      <td>9221</td>\n",
       "      <td>0.997091</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>78</td>\n",
       "      <td>74</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.74</td>\n",
       "      <td>5274</td>\n",
       "      <td>1.00</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>72</td>\n",
       "      <td>62</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.71</td>\n",
       "      <td>5274</td>\n",
       "      <td>0.999221</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>A2M</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.74</td>\n",
       "      <td>4844</td>\n",
       "      <td>1.00</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>58</td>\n",
       "      <td>50</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>4844</td>\n",
       "      <td>0.998054</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>99</td>\n",
       "      <td>17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.74</td>\n",
       "      <td>1023</td>\n",
       "      <td>0.00</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>84</td>\n",
       "      <td>62</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1023</td>\n",
       "      <td>0.751231</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tissue     gene  total_x  methylated_x  p_value_x  prior_x  length_x  \\\n",
       "0  Thymus     A1BG      130            83        1.0     0.74      1722   \n",
       "1  Thymus     A1CF       28            27        0.0     0.74      9221   \n",
       "2  Thymus    A2ML1       78            74        0.0     0.74      5274   \n",
       "3  Thymus      A2M       78            78        0.0     0.74      4844   \n",
       "4  Thymus  A3GALT2       99            17        1.0     0.74      1023   \n",
       "\n",
       "   p_value_low_x high_benign baseline_benign low_benign  total_y  \\\n",
       "0           0.01          no              no        yes      108   \n",
       "1           1.00         yes              no         no       24   \n",
       "2           1.00         yes              no         no       72   \n",
       "3           1.00         yes              no         no       58   \n",
       "4           0.00          no              no        yes       84   \n",
       "\n",
       "   methylated_y  p_value_y  prior_y  length_y  p_value_low_y low baseline high  \n",
       "0            76       0.60     0.71      1722       0.478854  no      yes   no  \n",
       "1            22       0.02     0.71      9221       0.997091  no       no  yes  \n",
       "2            62       0.00     0.71      5274       0.999221  no       no  yes  \n",
       "3            50       0.01     0.71      4844       0.998054  no       no  yes  \n",
       "4            62       0.33     0.71      1023       0.751231  no      yes   no  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#now conduct cds differential methylation analysis\n",
    "merged.head()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 15073 entries, 0 to 17881\n",
      "Data columns (total 20 columns):\n",
      "tissue             15073 non-null object\n",
      "gene               15073 non-null object\n",
      "total_x            15073 non-null int64\n",
      "methylated_x       15073 non-null int64\n",
      "p_value_x          15073 non-null float64\n",
      "prior_x            15073 non-null float64\n",
      "length_x           15073 non-null int64\n",
      "p_value_low_x      15073 non-null float64\n",
      "high_benign        15073 non-null object\n",
      "baseline_benign    15073 non-null object\n",
      "low_benign         15073 non-null object\n",
      "total_y            15073 non-null int64\n",
      "methylated_y       15073 non-null int64\n",
      "p_value_y          15073 non-null float64\n",
      "prior_y            15073 non-null float64\n",
      "length_y           15073 non-null int64\n",
      "p_value_low_y      15073 non-null float64\n",
      "low                15073 non-null object\n",
      "baseline           15073 non-null object\n",
      "high               15073 non-null object\n",
      "dtypes: float64(6), int64(6), object(8)\n",
      "memory usage: 2.4+ MB\n"
     ]
    }
   ],
   "source": [
    "#filter for genes that have at least x CpG's in them\n",
    "m = merged[(merged['total_x'] >= 10) & (merged['total_y'] >= 10)]\n",
    "m.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#9 possible transitions\n",
    "baseToBase = set([])\n",
    "baseToHigh = set([])\n",
    "baseToLow = set([])\n",
    "\n",
    "lowToHigh = set([])\n",
    "lowToBase = set([])\n",
    "lowToLow = set([])\n",
    "\n",
    "highToLow = set([])\n",
    "highToBase = set([])\n",
    "highToHigh = set([])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in m.iterrows():\n",
    "    if row['baseline_benign'] == 'yes' and row['baseline'] == 'yes':\n",
    "        baseToBase.add(row['gene'])\n",
    "    elif row['baseline_benign'] == 'yes' and row['high'] == 'yes':\n",
    "        baseToHigh.add(row['gene'])\n",
    "    elif row['baseline_benign'] == 'yes' and row['low'] == 'yes':\n",
    "        baseToLow.add(row['gene'])\n",
    "    elif row['low_benign'] == 'yes' and row['high'] == 'yes':\n",
    "        lowToHigh.add(row['gene'])\n",
    "    elif row['low_benign'] == 'yes' and row['baseline'] == 'yes':\n",
    "        lowToBase.add(row['gene']) \n",
    "    elif row['low_benign'] == 'yes' and row['low'] == 'yes':\n",
    "        lowToLow.add(row['gene'])\n",
    "    elif row['high_benign'] == 'yes' and row['low'] == 'yes':\n",
    "        highToLow.add(row['gene'])\n",
    "    elif row['high_benign'] == 'yes' and row['baseline'] == 'yes':\n",
    "        highToBase.add(row['gene'])\n",
    "    elif row['high_benign'] == 'yes' and row['high'] == 'yes':\n",
    "        highToHigh.add(row['gene'])\n",
    "    else:\n",
    "        print(row)\n",
    "    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2130\n",
      "289\n",
      "526\n",
      "112\n",
      "392\n",
      "3717\n",
      "98\n",
      "1997\n",
      "5812\n"
     ]
    }
   ],
   "source": [
    "print (len(baseToBase))\n",
    "print (len(baseToHigh))\n",
    "print (len(baseToLow))\n",
    "\n",
    "print (len(lowToHigh))\n",
    "print (len(lowToBase))\n",
    "print (len(lowToLow))\n",
    "\n",
    "print (len(highToLow))\n",
    "print (len(highToBase))\n",
    "print (len(highToHigh))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "oncogenes = m[(m['baseline_benign'] == 'yes') & (m['high_benign_y'] == 'yes')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oncogenes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import csv\n",
    "with open('lowToHigh.txt', 'w') as csvout:\n",
    "    writer = csv.writer(csvout, delimiter='\\n')\n",
    "    writer.writerow(list(lowToHigh))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# lowToHigh_df = m[m['gene'].isin(lowToHigh)]\n",
    "lowToHigh_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez GeneId</th>\n",
       "      <th>Genome Location</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Hallmark</th>\n",
       "      <th>Chr Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour Types(Somatic)</th>\n",
       "      <th>Tumour Types(Germline)</th>\n",
       "      <th>Cancer Syndrome</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Molecular Genetics</th>\n",
       "      <th>Role in Cancer</th>\n",
       "      <th>Mutation Types</th>\n",
       "      <th>Translocation Partner</th>\n",
       "      <th>Other Germline Mut</th>\n",
       "      <th>Other Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>29974</td>\n",
       "      <td>10:50806729-50859940</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>melanoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29974,A1CF,ENSG00000148584,Q9NQ94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>abl-interactor 1</td>\n",
       "      <td>10006</td>\n",
       "      <td>10:26748570-26860863</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>10p11.2</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>TSG, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10006,ABI1,ENSG00000136754,Q8IZP0,SSH3BP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene h...</td>\n",
       "      <td>25</td>\n",
       "      <td>9:130835447-130885683</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>9q34.1</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CML, ALL, T-ALL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T, Mis</td>\n",
       "      <td>BCR, ETV6, NUP214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25,ABL,ABL1,ENSG00000097007,P00519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL2</td>\n",
       "      <td>c-abl oncogene 2, non-receptor tyrosine kinase</td>\n",
       "      <td>27</td>\n",
       "      <td>1:179107718-179143044</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1q24-q25</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>ETV6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27,ABL2,ABLL,ENSG00000143322,P42684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACKR3</td>\n",
       "      <td>atypical chemokine receptor 3</td>\n",
       "      <td>57007</td>\n",
       "      <td>2:236580466-236581554</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2q37.3</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lipoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>HMGA2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A8K6J4,ACKR3,CMKOR1,CXCR7,ENSG00000144476,P251...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol                                               Name  \\\n",
       "0        A1CF                     APOBEC1 complementation factor   \n",
       "1        ABI1                                   abl-interactor 1   \n",
       "2        ABL1  v-abl Abelson murine leukemia viral oncogene h...   \n",
       "3        ABL2     c-abl oncogene 2, non-receptor tyrosine kinase   \n",
       "4       ACKR3                      atypical chemokine receptor 3   \n",
       "\n",
       "   Entrez GeneId        Genome Location  Tier Hallmark  Chr Band Somatic  \\\n",
       "0          29974   10:50806729-50859940     2      NaN  10q11.23     yes   \n",
       "1          10006   10:26748570-26860863     1      Yes   10p11.2     yes   \n",
       "2             25  9:130835447-130885683     1      Yes    9q34.1     yes   \n",
       "3             27  1:179107718-179143044     1      NaN  1q24-q25     yes   \n",
       "4          57007  2:236580466-236581554     1      Yes    2q37.3     yes   \n",
       "\n",
       "  Germline Tumour Types(Somatic) Tumour Types(Germline) Cancer Syndrome  \\\n",
       "0      NaN              melanoma                    NaN             NaN   \n",
       "1      NaN                   AML                    NaN             NaN   \n",
       "2      NaN       CML, ALL, T-ALL                    NaN             NaN   \n",
       "3      NaN                   AML                    NaN             NaN   \n",
       "4      NaN                lipoma                    NaN             NaN   \n",
       "\n",
       "  Tissue Type Molecular Genetics    Role in Cancer Mutation Types  \\\n",
       "0           E                NaN          oncogene            Mis   \n",
       "1           L                Dom       TSG, fusion              T   \n",
       "2           L                Dom  oncogene, fusion         T, Mis   \n",
       "3           L                Dom  oncogene, fusion              T   \n",
       "4           M                Dom  oncogene, fusion              T   \n",
       "\n",
       "  Translocation Partner Other Germline Mut Other Syndrome  \\\n",
       "0                   NaN                NaN            NaN   \n",
       "1                 KMT2A                NaN            NaN   \n",
       "2     BCR, ETV6, NUP214                NaN            NaN   \n",
       "3                  ETV6                NaN            NaN   \n",
       "4                 HMGA2                NaN            NaN   \n",
       "\n",
       "                                            Synonyms  \n",
       "0                  29974,A1CF,ENSG00000148584,Q9NQ94  \n",
       "1          10006,ABI1,ENSG00000136754,Q8IZP0,SSH3BP1  \n",
       "2                 25,ABL,ABL1,ENSG00000097007,P00519  \n",
       "3                27,ABL2,ABLL,ENSG00000143322,P42684  \n",
       "4  A8K6J4,ACKR3,CMKOR1,CXCR7,ENSG00000144476,P251...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "census = pd.read_csv('/Users/khandekara2/Documents/methylationProject/01_data/Census_all_2018.csv')\n",
    "census.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_genes = list(census['Gene Symbol'])\n",
    "baseToBase_cancer = set([])\n",
    "baseToHigh_cancer = set([])\n",
    "baseToLow_cancer = set([])\n",
    "\n",
    "lowToHigh_cancer = set([])\n",
    "lowToBase_cancer = set([])\n",
    "lowToLow_cancer = set([])\n",
    "\n",
    "highToLow_cancer = set([])\n",
    "highToBase_cancer = set([])\n",
    "highToHigh_cancer = set([])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in m.iterrows():\n",
    "    if row['baseline_benign'] == 'yes' and row['baseline'] == 'yes' and row['gene'] in cancer_genes:\n",
    "        baseToBase_cancer.add(row['gene'])\n",
    "    elif row['baseline_benign'] == 'yes' and row['high'] == 'yes' and row['gene'] in cancer_genes:\n",
    "        baseToHigh_cancer.add(row['gene'])\n",
    "    elif row['baseline_benign'] == 'yes' and row['low'] == 'yes' and row['gene'] in cancer_genes:\n",
    "        baseToLow_cancer.add(row['gene'])\n",
    "    elif row['low_benign'] == 'yes' and row['high'] == 'yes' and row['gene'] in cancer_genes:\n",
    "        lowToHigh_cancer.add(row['gene'])\n",
    "    elif row['low_benign'] == 'yes' and row['baseline'] == 'yes' and row['gene'] in cancer_genes:\n",
    "        lowToBase_cancer.add(row['gene']) \n",
    "    elif row['low_benign'] == 'yes' and row['low'] == 'yes' and row['gene'] in cancer_genes:\n",
    "        lowToLow_cancer.add(row['gene'])\n",
    "    elif row['high_benign'] == 'yes' and row['low'] == 'yes' and row['gene'] in cancer_genes:\n",
    "        highToLow_cancer.add(row['gene'])\n",
    "    elif row['high_benign'] == 'yes' and row['baseline'] == 'yes' and row['gene'] in cancer_genes:\n",
    "        highToBase_cancer.add(row['gene'])\n",
    "    elif row['high_benign'] == 'yes' and row['high'] == 'yes' and row['gene'] in cancer_genes:\n",
    "        highToHighbaseToBase = set([])\n",
    "    else:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "70\n",
      "9\n",
      "28\n",
      "2\n",
      "17\n",
      "122\n",
      "7\n",
      "63\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "print (len(baseToBase_cancer))\n",
    "print (len(baseToHigh_cancer))\n",
    "print (len(baseToLow_cancer))\n",
    "\n",
    "print (len(lowToHigh_cancer))\n",
    "print (len(lowToBase_cancer))\n",
    "print (len(lowToLow_cancer))\n",
    "\n",
    "print (len(highToLow_cancer))\n",
    "print (len(highToBase_cancer))\n",
    "print (len(highToHigh_cancer))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'NR4A3', 'CTNND2'}\n"
     ]
    }
   ],
   "source": [
    "print (lowToHigh_cancer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>gene</th>\n",
       "      <th>total_x</th>\n",
       "      <th>methylated_x</th>\n",
       "      <th>intermediate_x</th>\n",
       "      <th>p_value_x</th>\n",
       "      <th>prior_x</th>\n",
       "      <th>length_x</th>\n",
       "      <th>p_value_low_x</th>\n",
       "      <th>high_benign</th>\n",
       "      <th>...</th>\n",
       "      <th>total_y</th>\n",
       "      <th>methylated_y</th>\n",
       "      <th>intermediate_y</th>\n",
       "      <th>p_value_y</th>\n",
       "      <th>prior_y</th>\n",
       "      <th>length_y</th>\n",
       "      <th>p_value_low_y</th>\n",
       "      <th>low</th>\n",
       "      <th>baseline</th>\n",
       "      <th>high</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1423</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>PAX7</td>\n",
       "      <td>72</td>\n",
       "      <td>36</td>\n",
       "      <td>19</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.72</td>\n",
       "      <td>2260</td>\n",
       "      <td>0.00</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>29</td>\n",
       "      <td>24</td>\n",
       "      <td>73</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.65</td>\n",
       "      <td>2260</td>\n",
       "      <td>0.989859</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1688</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>PRRX1</td>\n",
       "      <td>40</td>\n",
       "      <td>19</td>\n",
       "      <td>3</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.72</td>\n",
       "      <td>2217</td>\n",
       "      <td>0.00</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>37</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.65</td>\n",
       "      <td>2217</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4058</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>TBX3</td>\n",
       "      <td>102</td>\n",
       "      <td>19</td>\n",
       "      <td>63</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.72</td>\n",
       "      <td>4723</td>\n",
       "      <td>0.00</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>54</td>\n",
       "      <td>52</td>\n",
       "      <td>138</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.65</td>\n",
       "      <td>4723</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9588</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>ALK</td>\n",
       "      <td>164</td>\n",
       "      <td>105</td>\n",
       "      <td>15</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.72</td>\n",
       "      <td>6220</td>\n",
       "      <td>0.02</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>58</td>\n",
       "      <td>47</td>\n",
       "      <td>154</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.65</td>\n",
       "      <td>6220</td>\n",
       "      <td>0.997658</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10000</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>CTNNA2</td>\n",
       "      <td>211</td>\n",
       "      <td>100</td>\n",
       "      <td>4</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.72</td>\n",
       "      <td>4349</td>\n",
       "      <td>0.00</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>69</td>\n",
       "      <td>68</td>\n",
       "      <td>202</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.65</td>\n",
       "      <td>4349</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12316</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>WWTR1</td>\n",
       "      <td>39</td>\n",
       "      <td>17</td>\n",
       "      <td>20</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.72</td>\n",
       "      <td>5030</td>\n",
       "      <td>0.00</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>55</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.65</td>\n",
       "      <td>5030</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12778</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>FAT4</td>\n",
       "      <td>287</td>\n",
       "      <td>174</td>\n",
       "      <td>10</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.72</td>\n",
       "      <td>16123</td>\n",
       "      <td>0.00</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>71</td>\n",
       "      <td>63</td>\n",
       "      <td>231</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.65</td>\n",
       "      <td>16123</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13492</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>CTNND2</td>\n",
       "      <td>196</td>\n",
       "      <td>118</td>\n",
       "      <td>10</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.72</td>\n",
       "      <td>5481</td>\n",
       "      <td>0.00</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>90</td>\n",
       "      <td>83</td>\n",
       "      <td>159</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.65</td>\n",
       "      <td>5481</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15878</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>FKBP9</td>\n",
       "      <td>66</td>\n",
       "      <td>33</td>\n",
       "      <td>6</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.72</td>\n",
       "      <td>3462</td>\n",
       "      <td>0.00</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>24</td>\n",
       "      <td>63</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.65</td>\n",
       "      <td>3462</td>\n",
       "      <td>0.996320</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17111</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>NR4A3</td>\n",
       "      <td>122</td>\n",
       "      <td>18</td>\n",
       "      <td>66</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.72</td>\n",
       "      <td>2588</td>\n",
       "      <td>0.00</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>90</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.65</td>\n",
       "      <td>2588</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       tissue    gene  total_x  methylated_x  intermediate_x  p_value_x  \\\n",
       "1423   Thymus    PAX7       72            36              19       1.00   \n",
       "1688   Thymus   PRRX1       40            19               3       1.00   \n",
       "4058   Thymus    TBX3      102            19              63       1.00   \n",
       "9588   Thymus     ALK      164           105              15       0.99   \n",
       "10000  Thymus  CTNNA2      211           100               4       1.00   \n",
       "12316  Thymus   WWTR1       39            17              20       1.00   \n",
       "12778  Thymus    FAT4      287           174              10       1.00   \n",
       "13492  Thymus  CTNND2      196           118              10       1.00   \n",
       "15878  Thymus   FKBP9       66            33               6       1.00   \n",
       "17111  Thymus   NR4A3      122            18              66       1.00   \n",
       "\n",
       "       prior_x  length_x  p_value_low_x high_benign ...  total_y methylated_y  \\\n",
       "1423      0.72      2260           0.00          no ...       29           24   \n",
       "1688      0.72      2217           0.00          no ...       11           11   \n",
       "4058      0.72      4723           0.00          no ...       54           52   \n",
       "9588      0.72      6220           0.02          no ...       58           47   \n",
       "10000     0.72      4349           0.00          no ...       69           68   \n",
       "12316     0.72      5030           0.00          no ...       11           11   \n",
       "12778     0.72     16123           0.00          no ...       71           63   \n",
       "13492     0.72      5481           0.00          no ...       90           83   \n",
       "15878     0.72      3462           0.00          no ...       28           24   \n",
       "17111     0.72      2588           0.00          no ...       14           14   \n",
       "\n",
       "       intermediate_y  p_value_y  prior_y  length_y  p_value_low_y low  \\\n",
       "1423               73       0.03     0.65      2260       0.989859  no   \n",
       "1688               37       0.01     0.65      2217       1.000000  no   \n",
       "4058              138       0.00     0.65      4723       1.000000  no   \n",
       "9588              154       0.01     0.65      6220       0.997658  no   \n",
       "10000             202       0.00     0.65      4349       1.000000  no   \n",
       "12316              55       0.01     0.65      5030       1.000000  no   \n",
       "12778             231       0.00     0.65     16123       0.999999  no   \n",
       "13492             159       0.00     0.65      5481       1.000000  no   \n",
       "15878              63       0.01     0.65      3462       0.996320  no   \n",
       "17111              90       0.00     0.65      2588       1.000000  no   \n",
       "\n",
       "       baseline high  \n",
       "1423         no  yes  \n",
       "1688         no  yes  \n",
       "4058         no  yes  \n",
       "9588         no  yes  \n",
       "10000        no  yes  \n",
       "12316        no  yes  \n",
       "12778        no  yes  \n",
       "13492        no  yes  \n",
       "15878        no  yes  \n",
       "17111        no  yes  \n",
       "\n",
       "[10 rows x 22 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lowToHigh_cancer_df = m[m['gene'].isin(lowToHigh_cancer)]\n",
    "lowToHigh_cancer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "{'WWTR1', 'PAX7', 'NR4A3', 'FKBP9', 'CTNNA2', 'PRRX1', 'FAT4', 'CTNND2', 'TBX3', 'ALK'}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
